### DISCUSSION TITLE
setup/config in lvim.plugins block doesn't work

### DISCUSSION THREAD

**Discussion by c02y** [LunarVim/LunarVim#2842] [Q&A]
I have this issue for a long time, when I add third-party plugins into `lvim.plugins` block in my `config.lua` file, the `function/setup` part never works, I'm following examples from https://www.lunarvim.org/plugins/02-extra-plugins.html , for example：
```
lvim.plugins = {
	{
		"rmagatti/goto-preview",
		config = function()
			require("goto-preview").setup({
				default_mappings = true,
			})
		end,
	},
}
```
Doesn't work, there are no mappings from `default_mappings`,

but
```
lvim.plugins = {
	{"rmagatti/goto-preview"},
}
require("goto-preview").setup({
	default_mappings = true,
})
```
Works very well, this kind of issue happens multiple times, I have to put the setup part out of the `lvim.plugins =` block to make it work.

Another example:
```
	{
		"rrethy/vim-hexokinase",
		run = "make hexokinase",
		config = function()
			vim.cmd("let g:Hexokinase_highlighters = ['backgroundfull']")
		end,
	},
```
`let g:Hexokinase_highlighters = ['backgroundfull']` doesn't work at all, I have to put this line into standalone `vim.cmd()` block which is not part of `lvim.plugins` block.

So is this a bug or I didn't use it right?

**Comment by cvalarida** [comment]
You're not alone! I'm running into the same problem. Using the following doesn't work:
```lua
lvim.plugins = {
  -- Other plugins omitted
  {
    "ggandor/leap.nvim",
    config = function()
      require('leap').set_default_keymaps()
    end
  }
}
```
The plugin loads, since I can `:help leap` and get the right documentation, and when I run `:lua require('leap').set_default_keymaps()`, it does indeed set the keymaps.

Moving that `require` line down below the `lvim.plugins` block like you did also worked for me, but this seems like a bug.

**Versions**
nvim v0.7.2
LunarVim on branch `master`, up-to-date at commit b04aefb

**Note:** This is probably better served as an issue rather than a discussion, though.

**Reply by rebuilt** [reply]
I tested the leap configuration posted above.  It works for me.  

Packer does not directly read your configuration file when running.  It uses an intermediary compilation step.  The packer file can be found in `~/.config/lvim/plugin/packer_compiled.lua`.  When making edits to a configuration, make sure you run `:PackerCompile` to regenerate the compilation file.

There are some cases where you really do need to place your configuration outside of the packer config function.   For example:

```lua
-- Can not be placed into the config method of the plugins.
lvim.builtin.cmp.formatting.source_names["copilot"] = "(Copilot)"
table.insert(lvim.builtin.cmp.sources, 1, { name = "copilot" })
```

**Reply by c02y** [reply]
> I tested the leap configuration posted above. It works for me.
> 
> Packer does not directly read your configuration file when running. It uses an intermediary compilation step. The packer file can be found in `~/.config/lvim/plugin/packer_compiled.lua`. When making edits to a configuration, make sure you run `:PackerCompile` to regenerate the compilation file.
> 
> There are some cases where you really do need to place your configuration outside of the packer config function. For example:
> 
> ```lua
> -- Can not be placed into the config method of the plugins.
> lvim.builtin.cmp.formatting.source_names["copilot"] = "(Copilot)"
> table.insert(lvim.builtin.cmp.sources, 1, { name = "copilot" })
> ```

Thanks, `:PackerCompile` works great, it solves my problem, I didn't know what  `Compile/Re-compile` in which-key window are for so long.

This is why I didn't post it as an issue.

Don't know why it must be called manually, is it possible to make it `Compile/Re-compile` automatically?

**Reply by cvalarida** [reply]
Ahhhhh, the intermediary compilation step is definitely what bit me. Thanks! 🙌 

+1 to the auto-compile idea. That'd be awesome.

**Reply by dpo** [reply]
`PackerCompile` doesn't solve the issue for me. I'm still stuck with what I described in https://github.com/LunarVim/LunarVim/discussions/2829. Configuring plugins outside of the `lvim.plugins` block does indeed work. What could cause this behavior?

My Lunarvim dashboard says
```
 lunarvim.org
 master-47e4e5b
```
That's the only version info I could find.

**Reply by c02y** [reply]
Maybe try to delete file `~/.config/lvim/plugin/packer_compiled.lua` and run `:PackerCompile` in new lvim session? Sometimes I found it doesn't work either by simply `:PackerCompile`

**Reply by rebuilt** [reply]
Something else to remember is that when you make changes to `~/.config/lvim/config.lua`, LunarVim will run a `:PackerSync`.  But if you configure your plugins in some external file, when you make changes to that file, LunarVim will not know about those changes and won't run `:PackerSync`.

**Reply by c02y** [reply]
> Something else to remember is that when you make changes to `~/.config/lvim/config.lua`, LunarVim will run a `:PackerSync`. But if you configure your plugins in some external file, when you make changes to that file, LunarVim will not know about those changes and won't run `:PackerSync`.

I don't think so, when I add a new plugin into `lvim.plugins` block, after I restart lvim, it won't install the plugin automatically, unless I manually execute `:PackerSync` or `:PackerInstall`

**Reply by danielo515** [reply]
> `PackerCompile` doesn't solve the issue for me. I'm still stuck with what I described in #2829. Configuring plugins outside of the `lvim.plugins` block does indeed work. What could cause this behavior?
> 
> My Lunarvim dashboard says
> 
> ```
>  lunarvim.org
>  master-47e4e5b
> ```
> 
> That's the only version info I could find.

You need packer clean, recompile and try several combinations of that
